### PR TITLE
#40 Add mount options parser and options validator

### PR DESCRIPTION
**Pull request by pzheng1025**

### CODE REVIEW

File: fuse_ll/src/fuse/mod.rs [hunk 1]
```diff
@@ -28,6 +28,7 @@ pub use request::Request;
 pub use session::Session;
 // pub use session::{Session, BackgroundSession};
 
+pub use mount::options_validator;
```
Comment (pwang7) [inline fuse_ll/src/fuse/mod.rs:31]: pub(crate) ?

File: fuse_ll/src/fuse/mount.rs [hunk 2]
```diff
@@ -8,23 +8,126 @@ use std::os::raw::c_void;
...
 use std::path::Path;
 
+use common_param::*;
 use param::*;
 
+/// Mount option validator function
+pub fn options_validator(option: String) -> Result<(), String> {
+    let ops = option.split(',');
+    for op in ops {
+        if !VALID_ON_OFF_OPTIONS.contains(&op) && !key_value_validator(&op) {
```
Comment (pwang7) [inline fuse_ll/src/fuse/mount.rs:18]: add comment about this if condition, what's the true case and what's the false case?

File: fuse_ll/src/fuse/mount.rs [hunk 3]
```diff
@@ -8,23 +8,126 @@ use std::os::raw::c_void;
...
+            return Err(format!("Invalid option \"{}\", valid options: ", op)
+                + &VALID_ON_OFF_OPTIONS.join(",")
+                + ","
+                + &VALID_KEY_VALUE_OPTIONS.join(","));
+        }
+    }
+    Ok(())
+}
+
+fn key_value_validator(option: &str) -> bool {
```
Comment (pwang7) [inline fuse_ll/src/fuse/mount.rs:28]: I think clap might do this kind of check

File: fuse_ll/src/fuse/mount.rs [hunk 4]
```diff
@@ -49,8 +153,12 @@ mod param {
     pub const MFSTYPENAMELEN: size_t = 16; // length of fs type name including null
     pub const MAXPATHLEN: size_t = 1024; //PATH_MAX
 
+    use super::common_param::*;
+    pub const VALID_ON_OFF_OPTIONS: [&str; 3] = [ALLOW_OTHER, RW, RO];
```
Comment (pwang7) [inline fuse_ll/src/fuse/mount.rs:157]: this is hard to maintain, what about use a function to define these options?

fn get_on_off_options() -> Vec<&str> {
    vec!["allow_other", "rw", "ro", ...]
}

File: fuse_ll/src/fuse/mount.rs [hunk 5]
```diff
@@ -64,6 +172,64 @@ mod param {
         pub random: u32,                      // random "secret" from device
         pub rdev: u32,                        // dev_t for the /dev/osxfuse{n} in question
     }
+
+    use log::error;
+    use std::ffi::CString;
+    impl FuseMountArgs {
+        fn add_option(options: Option<String>, option: &str) -> Option<String> {
```
Comment (pwang7) [inline fuse_ll/src/fuse/mount.rs:179]: maybe move this function to util mod?

File: fuse_ll/src/main.rs [hunk 6]
```diff
@@ -20,19 +20,18 @@ fn main() {
...
-            .map(|o| o.split(','))
-            .flatten()
-            .map(|o| OsStr::new(o))
-            .collect(),
+    let mountpoint = OsStr::new(matches.value_of("mountpoint").unwrap()); // safe to use unwrap() here, because mountpoint is required
+    let options: Vec<&str> = match matches.values_of("options") {
+        Some(options) => options.map(|o| o.split(',')).flatten().collect(),
         None => Vec::new(),
     };
+    dbg!(&options);
```
Comment (pwang7) [inline fuse_ll/src/main.rs:33]: remove dbg!(), or use debug!() to log.

File: fuse_ll/src/fuse/mount.rs [hunk 7]
```diff
@@ -8,23 +8,126 @@ use std::os::raw::c_void;
...
+    pub const RW: &str = "rw";
+    pub const RO: &str = "ro";
+    pub const FSNAME: &str = "fsname=<name>";
+
+    pub fn match_key_value(pattern: &str, option: &str) -> bool {
+        pattern.split('=').nth(0) == option.split('=').nth(0)
+    }
+
+    pub fn get_value(option: &str) -> String {
+        String::from(option.split('=').last().unwrap()) //Safe to use unwrap here, becuase option is always valid.
```
Comment (pwang7) [inline fuse_ll/src/fuse/mount.rs:56]: // safe to use unwrap here.
 ^ missing a space, may need cargo fmt before submitting PR

File: fuse_ll/src/fuse/mount.rs [hunk 8]
```diff
@@ -8,23 +8,126 @@ use std::os::raw::c_void;
...
+    false
+}
+
+mod common_param {
+    pub const ALLOW_OTHER: &str = "allow_other";
+    pub const RW: &str = "rw";
+    pub const RO: &str = "ro";
+    pub const FSNAME: &str = "fsname=<name>";
+
+    pub fn match_key_value(pattern: &str, option: &str) -> bool {
```
Comment (pwang7) [inline fuse_ll/src/fuse/mount.rs:51]: maybe leverage clap to handle this kind of match and comparison.

**Comment by pzheng1025** [comment]
Will submit a new PR.

**Pull request closed without merging**